src/backend/drm/gbm/surface.rs

```diff
@@ -0,0 +1,319 @@
+                    "Error releasing old front_buffer framebuffer: {:?}", err
+                );
+            }
+        }
+
+        // Drop the old surface after cleanup
+        *self.surface.borrow_mut() = surface;
+
+        Ok(())
+    }
+}
+
+impl<D: RawDevice + 'static> Surface for GbmSurface<D>
+where
+    <D as Device>::Return: ::std::borrow::Borrow<<D as RawDevice>::Surface>,
+{
+    type Connectors = <<D as Device>::Surface as Surface>::Connectors;
+    type Error = Error;
+
+    fn crtc(&self) -> crtc::Handle {
+        ::std::borrow::Borrow::borrow(&self.crtc).crtc()
+    }
+
+    fn current_connectors(&self) -> Self::Connectors {
+        ::std::borrow::Borrow::borrow(&self.crtc).current_connectors()
+    }
+
+    fn pending_connectors(&self) -> Self::Connectors {
+        ::std::borrow::Borrow::borrow(&self.crtc).pending_connectors()
+    }
+
+    fn add_connector(&self, connector: connector::Handle) -> Result<()> {
+        ::std::borrow::Borrow::borrow(&self.crtc)
+            .add_connector(connector)
+            .chain_err(|| ErrorKind::UnderlyingBackendError)
+    }
+
+    fn remove_connector(&self, connector: connector::Handle) -> Result<()> {
+        ::std::borrow::Borrow::borrow(&self.crtc)
+            .remove_connector(connector)
+            .chain_err(|| ErrorKind::UnderlyingBackendError)
+    }
+
+    fn current_mode(&self) -> Mode {
+        ::std::borrow::Borrow::borrow(&self.crtc).current_mode()
+    }
+
+    fn pending_mode(&self) -> Mode {
+        ::std::borrow::Borrow::borrow(&self.crtc).pending_mode()
+    }
+
+    fn use_mode(&self, mode: Mode) -> Result<()> {
+        ::std::borrow::Borrow::borrow(&self.crtc)
+            .use_mode(mode)
+            .chain_err(|| ErrorKind::UnderlyingBackendError)
+    }
+}
+
+// FIXME:
+//
+// Option 1: When there is GAT support, impl `GraphicsBackend` for `LegacyDrmBackend`
+//           using a new generic `B: Buffer` and use this:
+/*
+impl<'a, D: RawDevice + 'static> CursorBackend<'a> for GbmSurface<D>
+where
+    <D as Device>::Return: ::std::borrow::Borrow<<D as RawDevice>::Surface>,
+    <D as RawDevice>::Surface: CursorBackend<'a>,
+    <<D as RawDevice>::Surface as CursorBackend<'a>>::CursorFormat: Buffer,
+    <<D as RawDevice>::Surface as CursorBackend<'a>>::Error: ::std::error::Error + Send 
+{
+*/
+//
+// Option 2: When equality checks in where clauses are supported, we could at least do this:
+/*
+impl<'a, D: RawDevice + 'static> GraphicsBackend<'a> for GbmSurface<D>
+where
+    <D as Device>::Return: ::std::borrow::Borrow<<D as RawDevice>::Surface>,
+    <D as RawDevice>::Surface: CursorBackend<'a>,
+    <<D as RawDevice>::Surface as CursorBackend<'a>>::CursorFormat=&'a Buffer,
+    <<D as RawDevice>::Surface as CursorBackend<'a>>::Error: ::std::error::Error + Send 
+{
+*/
+// But for now got to do this:
+
+impl<'a, A: AsRawFd + 'static> CursorBackend<'a> for GbmSurface<LegacyDrmDevice<A>> {
+    type CursorFormat = &'a ImageBuffer<Rgba<u8>, Vec<u8>>;
```

Comment:
Does that mean that it will not be possible to efficiently handle clients that use opengl to draw their cursor images? (Tbh I have no idea if that can actually happen, but in general I'd think that the ability to use OpenGL on a surface should be orthogonal to its role)

Reply:
Actually this does hardware compositing (draw the cursor via a seperate drm plane instead of drawing it via GL in software), which should be way faster (and does not require a redraw of the GL-Surface to be visible). So this is (when possible) the fastest implementation.

Converting certain formats (like the xcursor) into an ImageBuffer *could* be slow, but that should really be a problem for very frequent cursor changes.

In generally I would like to allow drawing a software cursor via GL as well, but I believe I need the `specialization` feature to be able to provide multiple stacking implementations and do this right. Maybe we could also move `CursorFormat` from an associated type to generic parameters of the trait, so we can actually provide multiple implementations depending on the actual format.

Right now every surface does provide an implementation that should be as fast as possible (although not very flexible).
Except for `GbmSurface`, which does only provide an implementation for `GbmSurface<LegacyDrmDevice<A>>`, if you provide your own `RawDevice` you will need to implement this yourself (for now).

Reply:
Okay I see. So if a client decides to use OpenGL to draw its cursor image, the compositor will have to extract the `EGLImage` into an `ImageBuffer` to then composite it on the hardware, right? And the performance penalty should not be big (assuming the client doesn't set up ridiculously big cursor images)?

I agree this is a pretty niche question anyway, we can absolutely come back to it at a later time (or never if it's not needed).

Reply:
Well the client could also decide not to use the `CursorBackend` (or clear the cursor by providing a clear image (maybe we should provide an api for that)) and just render a software cursor itself.

That might actually be the best way to handle this case. Provide an implementation for hardware compositing and let clients do their own software cursor, if they want to.

Reply:
Yeah, lets forget about that for now, I'm just being needlessly nitpicking about hypothetical edge-case uses.